rbd: use clusterID from the volumeContext when handling nodestage request for migration

humblec opened the pull request:
previously we were retriving clusterID using the monitors field
in the volume context at node stage code path. however it is possible to
retrieve clusterID from the volumeContext passed with  nodeStage request

Additional Info:
https://github.com/ceph/ceph-csi/pull/2523#issuecomment-931130464
    
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -291,13 +273,19 @@ func (ns *NodeServer) NodeStageVolume(
+		if isMigrationVolID(volID) {
+			parsedMigHandle, pErr := parseMigrationVolID(volID)
+			if pErr != nil {
+				log.ErrorLog(ctx, "failed to parse the migration handle, err :%v ", pErr)
```
just use `failed to parse the migration handle: %v` as error format, the last component after `:` is expected to be the error message

humblec replied:
done

nixpanic commented on internal/rbd/migration_test.go:
```diff
@@ -171,3 +171,42 @@ func TestParseMigrationVolID(t *testing.T) {
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
```
I really hate that we're doing it this way. It is very ugly to overload a `tt` variable like this. Please don't introduce these kind of things for new tests, use a different variable instead.

humblec replied:
well, its just a variable name' ? as long as "memory" is concerned it is totally different location :). No hard opinion to change the variable name though. can be changed.

humblec commented:
/retest all

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -291,13 +273,19 @@ func (ns *NodeServer) NodeStageVolume(
+		if isMigrationVolID(volID) {
+			parsedMigHandle, pErr := parseMigrationVolID(volID)
+			if pErr != nil {
+				log.ErrorLog(ctx, "failed to parse the migration handle: %v ", pErr)
```
drop the trailing space at `%v "`

humblec replied:
done.

nixpanic reviewed (commented):
I don't have any concerns with this change. It is not clear to me why this is better or what bug it addresses. That could be explained a little more in the commit message.

humblec commented:
> I don't have any concerns with this change. It is not clear to me why this is better or what bug it addresses. That could be explained a little more in the commit message.

my bad, this is actually a small change which marked as a followup item https://github.com/ceph/ceph-csi/pull/2523#issuecomment-931130464.. Added this information in the PR description too. sorry for missing it in first place @nixpanic

humblec commented:
@Mergifyio rebase

humblec commented:
/retest all

humblec commented:
/retest ci/centos/k8s-e2e-external-storage/1.21

humblec commented:
@Mergifyio rebase

Madhu-1 reviewed (dismissed):
Please remove GetClusterIDFromMon and   readClusterInfoWithMon which is not required anymore https://github.com/ceph/ceph-csi/pull/2512/files#diff-b9c67ae3efee5ec6b2d1889405ffaf967a4808f3d9a12a287ac6faa9ba19d762R169-R207 .
Please also make required changes in E2E to remove monitor information from PV.CSI  object as  its not required anymore.

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -292,13 +274,19 @@ func (ns *NodeServer) NodeStageVolume(
+		if isMigrationVolID(volID) {
+			parsedMigHandle, pErr := parseMigrationVolID(volID)
+			if pErr != nil {
+				log.ErrorLog(ctx, "failed to parse the migration handle: %v", pErr)
```
`failed to parse the migration handle`  to `failed to parse volumeID`

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -292,13 +274,19 @@ func (ns *NodeServer) NodeStageVolume(
+				return nil, status.Error(codes.Internal, pErr.Error())
+			}
+			// store clusterID
+			req.VolumeContext[util.ClusterIDKey] = parsedMigHandle.clusterID
```
as you already know the clusterID can't we just add that to volumeContext in kubernetes itself?

humblec replied:
@Madhu-1 its possible on both ways, however kept this way as volID considered to be CSI  handler which can be depended for the operations.

Madhu-1 replied:
in nodeStage operation, we retrieve the clusterID from the context. IMO we should take the same route as it's possible now.

humblec replied:
@Madhu-1 if really needed I can do that, wdyt?  why I am in bit hesitation here is, it also need changes in the migration Context func ...etc. Can that be done later? I can mark it as todo for now . Is that fine ?

Madhu-1 replied:
instead of adding complexity am more inclined to do this as it's going to be minimal change.

humblec replied:
well, this patch submission was done to use mons hash as in the previous PR review comment, now the change request is actually different :).

Madhu-1 replied:
@humblec I do not always think about all the possibilities sometimes we suggesting things and when actually reviewing PR's sometimes I will try to think about the possibilities. if we get good approach IMO not a bad to change in between instead of changing later :)

humblec replied:
`@humblec I do not always think about all the possibilities sometimes we suggesting things and when actually reviewing PR's sometimes I will try to think about the possibilities. if we get good approach IMO not a bad to change in between instead of changing later :)` - hmmm..as mentioned both works, any way addressing this comment then.

Madhu-1 reviewed (commented):
@humblec sorry for  it late, i have a question PTAL

humblec commented:
> Please remove GetClusterIDFromMon and readClusterInfoWithMon which is not required anymore https://github.com/ceph/ceph-csi/pull/2512/files#diff-b9c67ae3efee5ec6b2d1889405ffaf967a4808f3d9a12a287ac6faa9ba19d762R169-R207 . Please also make required changes in E2E to remove monitor information from PV.CSI object as its not required anymore.

@Madhu-1 I kept that out there in case if we want to use the same in future , but I can remove that as well.  Done.

humblec commented:
/retest all

humblec commented:
`Error from server: error when creating "https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v4.0.0/client/config/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml": etcdserver: request timed out
script returned exit code 1`

humblec commented:
I have to adjust the commit message and little bit of splitting the last commit.. on it

humblec committed:
rbd: change iteration variable name in the migration test to be specific

we reuse or overload the variable name in the test execution at present.
This commit use a different variable name as initialized in each run

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
rbd: use clusterID from volumeContext at nodestage

previously we were retriving clusterID using the monitors field
in the volume context at node stage code path. however it is possible to
retrieve or use clusterID directly from the volume context. This
commit also remove the getClusterIDFromMigrationVolume() function
which was used previously and its tests

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
e2e: adjust migration tests to use clusterID in the volume context

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
> I have to adjust the commit message and little bit of splitting the last commit.. on it

Done..

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.20

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.20

humblec commented:
@Mergifyio rebase

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.20

humblec commented:
Can you please review this ?

yati1998 reviewed (commented):
LGTM. Will approve once any of the maintainers approve it.

Madhu-1 commented on internal/rbd/migration_test.go:
```diff
@@ -156,17 +156,56 @@ func TestParseMigrationVolID(t *testing.T) {
+		{
+			"nil clusterID set",
+			map[string]string{"migration": "true", "clusterID": ""},
+			true,
```
this test seems redundant as there is no clusteID validation in `isMigrationContext` can you please remove this test?

Madhu-1 commented on internal/rbd/nodeserver.go:
```diff
@@ -291,14 +273,14 @@ func (ns *NodeServer) NodeStageVolume(
+		if req.GetVolumeContext()[util.ClusterIDKey] == "" {
+			log.ErrorLog(ctx, "clusterID is nil in the volume context")
+
+			return nil, status.Error(codes.InvalidArgument, "missing required parameter clusterID")
```
IMO this check can be removed as the clusterID validation will be done when populating the rbdVolume just like other NodeStageRequest, we don't need to tread this as special

humblec replied:
Thought an extra check here is good for fail early, but removing it.

Madhu-1 reviewed (commented):
Currently, we have 3 e2e tests and we can move the duplicate test for Block and See if you can adjust other tests to validate Image deletion.

* validate RBD migration+static Block PVC Deletion
* validate RBD migration+static FileSystem PVC
* validate RBD migration+static Block PVC

humblec commented:
@Madhu-1 addressed the final comments. ptal.. thanks..

Madhu-1 commented on internal/rbd/migration.go:
```diff
@@ -32,6 +32,12 @@ func isMigrationVolID(volHash string) bool {
+// migration specific parameters.
+func isMigrationContext(vc map[string]string) bool {
+	return vc[intreeMigrationKey] == intreeMigrationLabel
+}
```
this is not required anymore and can be removed?

humblec replied:
@Madhu-1 done. ptal

humblec commented:
@Mergifyio rebase

humblec commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20
